### PR TITLE
Add iFlight Commando8

### PR DESCRIPTION
**Pull request by XING-IF**
Add iFlight Commando8 ELRS radio.
Based on the OpenTX PR #8905 with fixes.
https://github.com/opentx/opentx/pull/8905

[https://shop.iflight-rc.com/Commando-8-Radio-Transmitter-Pro1696](url)

image

### CODE REVIEW

File: companion/src/simulation/simulatorwidget.cpp
Comment (pfeerick) [inline companion/src/simulation/simulatorwidget.cpp:116]: Shouldn't this be `SimulatedUIWidgetCommando8`?

**Review by pfeerick — COMMENTED** [review]
This is looking a lot better than the original code I saw a few weeks ago :)

Because of the recent changes to the build system (#1736) in `/tools` will need updating.

File: companion/src/simulation/simulatorwidget.cpp
Comment (XING-IF) [reply companion/src/simulation/simulatorwidget.cpp]: oh thanks, fixed!

File: radio/src/targets/common/arm/stm32/audio_dac_driver.cpp
Comment (pfeerick) [inline radio/src/targets/common/arm/stm32/audio_dac_driver.cpp:63]: This is a common header, used by colorlcd and B&W radios... however you've only defined these `setMutePin, resetMutePin, readMutePinLevel` you've only defined them for the taranis in `board.cpp`. As to what the preferred resolution is, this is up to @raphaelcoeffic to decide. However, this is what is breaking the build checks for this PR currently as far as I can tell.

Comment (XING-IF) [reply radio/src/targets/common/arm/stm32/audio_dac_driver.cpp:63]: Yes, this part is an open problem we were having... Hope we could get some feedback for a better solution

File: .github/workflows/actions.yml
Comment (raphaelcoeffic) [inline .github/workflows/actions.yml:41]: It should be sufficient to the target to the "build-only" actions (bellow), as it does not change anything fundamentally.

File: .github/workflows/actions.yml
Comment (raphaelcoeffic) [inline .github/workflows/actions.yml:86]: This one can stay.

File: companion/src/firmwares/opentx/opentxinterface.cpp
Comment (raphaelcoeffic) [inline companion/src/firmwares/opentx/opentxinterface.cpp:1477]: Please be aware that the compilation options are not functional for now. The 3 options listed here are also on the list of options that will probably be removed soon (selected only at run-time).

File: radio/src/gui/128x64/model_setup.cpp
Comment (raphaelcoeffic) [inline radio/src/gui/128x64/model_setup.cpp:1009]: As I understand it, you are trying to implement mutually exclusive modules between internal and external, right?
This should be implemented in `isInternalModuleAvailable` and `isExternalModuleAvailable`.

Also, the `define` should be called something like `MUTUALLY_EXCLUSIVE_MODULES`.

File: radio/src/gui/128x64/model_setup.cpp
Comment (raphaelcoeffic) [inline radio/src/gui/128x64/model_setup.cpp:1023]: See previous comment. Implement this in `isExternalModuleAvailable`.

File: radio/src/opentx.cpp
Comment (raphaelcoeffic) [inline radio/src/opentx.cpp:361]: Please use `IFLIGHT_RELEASE` here. We should not have different defaults based on the radio. Using the proper tag allows you however to have the se defaults in your own builds & releases.

File: radio/src/pulses/modules_helpers.h
Comment (raphaelcoeffic) [inline radio/src/pulses/modules_helpers.h:845]: What is the purpose of this? this looks like a hack. Please explain.

File: radio/src/sdcard.h
Comment (raphaelcoeffic) [inline radio/src/sdcard.h:166]: New radios do not need this, as this is only for binary storage.

File: radio/src/storage/conversions/datastructs_219.h
Comment (raphaelcoeffic) [inline radio/src/storage/conversions/datastructs_219.h:31]: Please remove all changes in `storage/conversions/*`: you would only need them to support a pre-existing binary storage. As EdgeTx does not support binary storage for new radios, it should not be needed.

File: radio/src/storage/conversions/yaml/dataconstants_220.h
Comment (raphaelcoeffic) [inline radio/src/storage/conversions/yaml/dataconstants_220.h:61]: Please remove all changes in `storage/conversions/*`: you would only need them to support a pre-existing binary storage. As EdgeTx does not support binary storage for new radios, it should not be needed.

File: radio/src/storage/conversions/yaml/yaml_datastructs_commando8.cpp
Comment (raphaelcoeffic) [inline radio/src/storage/conversions/yaml/yaml_datastructs_commando8.cpp:7]: Please remove all changes in `storage/conversions/*`: you would only need them to support a pre-existing binary storage. As EdgeTx does not support binary storage for new radios, it should not be needed.

File: radio/src/storage/conversions/yaml/yaml_datastructs.cpp
Comment (raphaelcoeffic) [inline radio/src/storage/conversions/yaml/yaml_datastructs.cpp:58]: Please remove all changes in `storage/conversions/*`: you would only need them to support a pre-existing binary storage. As EdgeTx does not support binary storage for new radios, it should not be needed.

File: radio/src/targets/common/arm/stm32/audio_dac_driver.cpp
Comment (raphaelcoeffic) [reply radio/src/targets/common/arm/stm32/audio_dac_driver.cpp:63]: The better way would be to provide these functions in the driver itself, and use a specific `define` called `INVERTED_MUTE_PIN` in `hal.h` to express that the logic is inverted.

In `targets/taranis/hal.h`, you would then define `INVERTED_MUTE_PIN` only for your target.

File: radio/src/targets/common/arm/stm32/bootloader/boot.cpp
Comment (raphaelcoeffic) [inline radio/src/targets/common/arm/stm32/bootloader/boot.cpp:539]: See discussion here: https://github.com/opentx/opentx/pull/8905#discussion_r851485381

This code should be best replaced by a single function called `blExit()` which is implemented differently in the target's `bootmenu.cpp`. There you could place `#if defined(...)` and replace the default taranis implementation.

File: radio/src/targets/common/arm/stm32/pwr_driver.cpp
Comment (raphaelcoeffic) [inline radio/src/targets/common/arm/stm32/pwr_driver.cpp:52]: I'm trying to empty bit-by-bit this strange "pot pourri" driver. Please squeeze this into `boardInit()` for your target.

File: radio/src/targets/taranis/CMakeLists.txt
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/CMakeLists.txt:255]: It is probably preferable to have a parameter named `USE_RTC_CLOCK` defined at global level of this file, and force its value here.

File: radio/src/targets/taranis/CMakeLists.txt
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/CMakeLists.txt:541]: Use this instead (see previous comment):
```cmake
add_definitions(-DAUDIO -DVOICE)

if(USE_RTC_CLOCK)
  add_definitions(-DRTCLOCK)
endif()
```

File: radio/src/targets/taranis/board.cpp
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/board.cpp:447]: please move these functions to the driver itself and declare them `static` (private to the compile unit).

Comment (raphaelcoeffic) [reply radio/src/targets/taranis/board.cpp:447]: Also use only the state as parameter (for instance: `bool enabled`).

File: radio/src/targets/taranis/board.h
Comment (raphaelcoeffic) [inline radio/src/targets/taranis/board.h:830]: While moving the functions to the audio driver, please remove these as well.

File: tools/build-companion-nightly.sh
Comment (raphaelcoeffic) [inline tools/build-companion-nightly.sh:219]: This has been replaced by a loop over all the targets (rebase/merge issue?). Please fix.

**Review by raphaelcoeffic — CHANGES_REQUESTED** [review]

File: radio/src/pulses/modules_helpers.h
Comment (XING-IF) [reply radio/src/pulses/modules_helpers.h:845]: Because our D8 module crystal frequency is 26.000050MHZ, and the R81 receiver crystal frequency is 25.999340MHZ, so there is a frequency deviation and the frequency needs to be compensated. This variable g_model.moduleData[moduleIdx].multi.optionValue cannot save the setting value after switching to another module, so it is set to -27 by default. Is there a better approach for this?

File: radio/src/gui/128x64/model_setup.cpp
Comment (WONGZEONJYU) [reply radio/src/gui/128x64/model_setup.cpp:1009]: YES

File: radio/src/pulses/modules_helpers.h
Comment (raphaelcoeffic) [reply radio/src/pulses/modules_helpers.h:845]: Ouch! Ideally this would be fixed with a specific MPM build. I believe this can be set via a `define` in the MPM code. You could then provide specific MPM firmware builds to address this.

**Comment by XING-IF** [comment]
All suggestions have been implemented and code should be fixed. We kindly ask for a re-review or further suggestions! Thanks guys

File: radio/src/targets/taranis/board.cpp
Comment (pfeerick) [reply radio/src/targets/taranis/board.cpp]: I think this was resolved cleaner by removing the funcs and just doing the inversion based on the define

File: radio/src/targets/common/arm/stm32/bootloader/boot.cpp
Comment (pfeerick) [inline radio/src/targets/common/arm/stm32/bootloader/boot.cpp:532]: Since this is a common file... This needs either making conditional on the B&W radios, or implementations added for horus and nv14. After that this PR *should* start to build for all targets again. Since Raphael *did* mention blExit could be implemented differently for each target, I would suggest the latter.

```diff
diff --git a/radio/src/targets/horus/bootloader/boot_menu.cpp b/radio/src/targets/horus/bootloader/boot_menu.cpp
index 089bb3868..7a86981a9 100644
--- a/radio/src/targets/horus/bootloader/boot_menu.cpp
+++ b/radio/src/targets/horus/bootloader/boot_menu.cpp
@@ -235,3 +235,10 @@ void bootloaderDrawFilename(const char* str, uint8_t line, bool selected)
         lcd->drawSolidRect(119, 72 + (line * 25), 278, 26, 2, BL_SELECTED);
     }
 }
+
+void blExit()
+{
+      lcdClear();
+      lcdRefresh();
+      lcdRefreshWait();
+}
diff --git a/radio/src/targets/nv14/bootloader/boot_menu.cpp b/radio/src/targets/nv14/bootloader/boot_menu.cpp
index af6d49cee..4501dd034 100644
--- a/radio/src/targets/nv14/bootloader/boot_menu.cpp
+++ b/radio/src/targets/nv14/bootloader/boot_menu.cpp
@@ -240,3 +240,10 @@ void bootloaderDrawFilename(const char* str, uint8_t line, bool selected)
         lcd->drawSolidRect(DEFAULT_PADDING + 25, 72 + (line * 25), LCD_W - (DEFAULT_PADDING + 25) - 28, 26, 2, BL_SELECTED);
     }
 }
+
+void blExit()
+{
+      lcdClear();
+      lcdRefresh();
+      lcdRefreshWait();
+}
```

File: radio/src/pulses/modules_helpers.h
Comment (XING-IF) [reply radio/src/pulses/modules_helpers.h]: is this now implemented correctly?

File: radio/src/targets/common/arm/stm32/bootloader/boot.cpp
Comment (XING-IF) [reply radio/src/targets/common/arm/stm32/bootloader/boot.cpp:532]: ok thanks, then let's wait for his review

File: radio/src/gui/gui_common.cpp
Comment (raphaelcoeffic) [inline radio/src/gui/gui_common.cpp:647]: Why not using simply this?
```c++
#if defined(MUTUALLY_EXCLUSIVE_MODULES)
  if (!isModuleNone(EXTERNAL_MODULE)) return false;
#endif
```

File: radio/src/gui/gui_common.cpp
Comment (raphaelcoeffic) [inline radio/src/gui/gui_common.cpp:695]: #if defined(MUTUALLY_EXCLUSIVE_MODULES)
  if (!isModuleNone(INTERNAL_MODULE)) return false;
#endif

File: radio/src/pulses/modules_helpers.h
Comment (pfeerick) [reply radio/src/pulses/modules_helpers.h]: Are we talking about frequency tuning for the CC2500 here?? If so, that _normally_ needs to be done on a unit by unit basis anyway. But if you know there is a offset to start with, you can set the forced tuning defines with a build of the MPM firmware and set the default values at the MPM firmware level. https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/docs/Frequency_Tuning.md#forced-tuning-values

Comment (XING-IF) [reply radio/src/pulses/modules_helpers.h]: Yes, it's a CC2500 module. We'll checkout the link. thanks

File: radio/src/gui/gui_common.cpp
Comment (XING-IF) [reply radio/src/gui/gui_common.cpp]: fixed

File: radio/src/pulses/modules_helpers.h
Comment (XING-IF) [reply radio/src/pulses/modules_helpers.h]: We fixed the code of our module. No EdgeTX code adaption necessary anymore.

**Comment by pfeerick** [comment]
This is looking starting to look pretty good now. No more breaking of other builds and code is starting to look reasonably clean :)  

@raphaelcoeffic What needs to be done to kick the FreeRTOS change, as I don't think that should be there?

**Comment by XING-IF** [comment]
seems like all check passed, no broken builds neither

**Comment by raphaelcoeffic** [comment]
> This is looking starting to look pretty good now. No more breaking of other builds and code is starting to look reasonably clean :)
> 
> @raphaelcoeffic What needs to be done to kick the FreeRTOS change, as I don't think that should be there?

Looks like this slipped in: I just rebased on `main` and edited the culprit to remove the FreeRTOS change.

**Review by raphaelcoeffic — APPROVED** [review]

**Comment by pfeerick** [comment]
@elecpower As the Companion whisperer, I defer to you for the Companion side of this :)

**Comment by elecpower** [comment]
Compiled branch and after a very quick check there are basic issues. So for starters Companion requires updated/checked:
- board capabilities
- firmware capabilities
- isAvailable functions
- radio profile default module
- radio hardware settings
- radio to radio conversions
- reading opentx bin format
- write yaml format 
- simulator graphic as per @pfeerick above
- etc

**Comment by raphaelcoeffic** [comment]
@elecpower side note: there is no OpenTx binary format here. The radio is basically unsupported on OpenTx side.

**Comment by XING-IF** [comment]
> @elecpower side note: there is no OpenTx binary format here. The radio is basically unsupported on OpenTx side.

The OpenTX PR would be https://github.com/opentx/opentx/pull/8905 but is not yet through.

**Comment by deadbytefpv** [comment]
Switch Config doesn't seem to stick.
SA to SD goes back into "None" after each power-cycle.
Radio.YML have these set up though.

**Comment by pfeerick** [comment]
Mine arrived today, and I just flashed the latest PR build onto it. I initially got the 'None' setting for the switches (which I'm not so concerned about yet - will have to see what happens on a clean config - as it should be defaulting to correct setting if invalid), and then after setting them, it has stuck for several reboot cycles. Are you changing pages or exiting the radio setup menu before powering off the radio after setting them up?

**Comment by deadbytefpv** [comment]
> Mine arrived today, and I just flashed the latest PR build onto it. I initially got the 'None' setting for the switches (which I'm not so concerned about yet - will have to see what happens on a clean config - as it should be defaulting to correct setting if invalid), and then after setting them, it has stuck for several reboot cycles. Are you changing pages or exiting the radio setup menu before powering off the radio after setting them up?

Yes. I've actually kept the radio turned quite a while before turning it off to see if the settings will stick.. 
I've also tested with a clean sd card to perhaps refresh the configs, but it still happened (settings went back to None after a reboot).
Thanks for checking things out!

**Comment by XING-IF** [comment]
@pfeerick @deadbytefpv I just checked with iFlight. The radio was obviously shipped with a custom build from their drives which is different to the code of this PR. The firmware version is misleading and says "Built by EdgeTX" which should actually say "Built by iFlight" as of this point the code was not merged yet. There were some fixes on their drive which have not been provided for this PR. I'm just trying to get some answers who fu**ed up again...

**Comment by pfeerick** [comment]
As far as the build it was shipped with, it seemed OK as far as the switches, but I didn't really try anything with it as I expected it to be out of date. Certainly need to liven someone ul if they're not at least using the current state of this PR as the shipping build 😅

**Comment by XING-IF** [comment]
> As far as the build it was shipped with, it seemed OK as far as the switches, but I didn't really try anything with it as I expected it to be out of date. Certainly need to liven someone ul if they're not at least using the current state of this PR as the shipping build 😅

I'm already releasing my anger and making sure this is a learned lesson. Mistakes can happen but should actually be found where they are made!

**Comment by deadbytefpv** [comment]
I had a little bit more play...
Added a second model from scratch. 
Reconfigured the switches and also setup my mixes with those switches.. 
Also did the mixes on the first model (that came with the radio).
Let it sit a bit.. some 3 minutes or so.
Turned it off and didn't touched the radio for a while (turned off).

Just turned it On a moment ago and the switches are there. Didn't reset to None. On both models.
I've since deleted the second model. Switched to the first original model. I've turned the radio off, and powered it back on, and the switches seems to be intact.
I'll have some more play later and see if switch config sticks.
Will report here if there's any change.
Thanks!

**Comment by raphaelcoeffic** [comment]
Obsoleted by #1982

**Pull request closed without merging**
